### PR TITLE
Processing for a lo/l1b/goodtimes product

### PR DESCRIPTION
**Pull request by vineetbansal**
# Change Summary

Changed the `goodtimes` Lo/l1b product algorithm. Both the algorithm as well as the structure of the output datasets have changed. I've run the algorithm by the instrument team. Since no other products depend on these, this should be okay.

## File changes

`l1b_bgrates_and_goodtimes` function in `imap_processing/lo/l1b/lo_l1b.py` changed to adhere to new algorithm from IMAP-Lo team (not yet officially documented). Tests related to this function `l1b_bgrates_and_goodtimes` have been tweaked.

## Testing

Modified tests to test some edge cases for the new algorithm. Not all edge cases have been tested, but I'll add more in a subsequent commit.

### CODE REVIEW

File: imap_processing/lo/l1b/lo_l1b.py [hunk 1]
```diff
@@ -2831,8 +2848,372 @@ def l1b_bgrates_and_goodtimes(  # noqa: PLR0912
...
+
+    epoch_ttj2000 = cdf_hist["epoch"].values
+    n_epochs = epoch_ttj2000.shape[0]
+    met = ttj2000ns_to_met(epoch_ttj2000)
+
+    # Get year and day-of-year for the anti-RAM threshold override lookup
+    epoch_utc_str = et_to_utc(ttj2000ns_to_et(epoch_ttj2000[0]))
+    epoch_start_dt = datetime.strptime(epoch_utc_str.split("T")[0], "%Y-%m-%d")
+    epoch_year = epoch_start_dt.year
+    epoch_doy = epoch_start_dt.timetuple().tm_yday
```
Comment (tmplummer) [inline imap_processing/lo/l1b/lo_l1b.py]: Use `time.epoch_to_doy` instead.

File: imap_processing/lo/l1b/lo_l1b.py [hunk 2]
```diff
@@ -2831,8 +2848,372 @@ def l1b_bgrates_and_goodtimes(  # noqa: PLR0912
...
+    pivot: float = 0.0
+
+    if "pcc_coarse_pot_pri" in cdf_hk:
+        hk_epoch_ets = ttj2000ns_to_et(cdf_hk["epoch"])
+        start_et_hk = (
+            hk_epoch_ets[0] + timedelta(hours=PIVOT_HK_HOUR_RANGE[0]).total_seconds()
+        )
+        end_et_hk = (
+            hk_epoch_ets[0] + timedelta(hours=PIVOT_HK_HOUR_RANGE[1]).total_seconds()
+        )
```
Comment (tmplummer) [inline imap_processing/lo/l1b/lo_l1b.py:2557]: Fragile code. Nominal pointings are close to 24 hours, but not all pointings are that long. For example around the TCM on April 7, the repoint table has the following:
```
513165615,0,513165816,0,2026-04-06 10:00:00.748931,2026-04-06 10:03:21.748756,208
513252017,0,513252218,0,2026-04-07 10:00:02.673645,2026-04-07 10:03:23.673469,209
513287744,0,513287946,0,2026-04-07 19:55:29.642514,2026-04-07 19:58:51.642338,210
513288856,0,513289642,0,2026-04-07 20:14:01.641545,2026-04-07 20:27:07.640860,211
513338414,0,513338796,0,2026-04-08 09:59:59.598459,2026-04-08 10:06:21.598127,212
```
Pointing 209 is ~9:52 in duration, pointing 210 is ~0:16 in duration, etc.

File: imap_processing/lo/constants.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+# Hours into the day (UTC) for HK data to calculate median for pivot angle estimation.
+PIVOT_HK_HOUR_RANGE: tuple[int, int] = (3, 15)
+
+# Per-day overrides for the anti-RAM background rate. Keyed by (year, day-of-year)
+BG_RATE_ANTI_RAM_OVERRIDES: dict[tuple[int, int], float] = {
+    (2026, 62): 0.0014,
+    (2026, 64): 0.0,
+    (2026, 65): 0.0,
+    (2026, 91): 0.03,
+}
```
Comment (tmplummer) [inline imap_processing/lo/constants.py]: Hard coded override days 😬. Two major concerns with this:
1. This required code updates and build/deploy of the Lo container each time an update to these is desired.
2. Per the PI, the pipeline should be fully automated with no manual intervention to tweak algorithm parameters.

File: imap_processing/lo/l1b/lo_l1b.py [hunk 4]
```diff
@@ -2515,319 +2511,295 @@ def l1b_bgrates_and_goodtimes(  # noqa: PLR0912
...
+        elem_anti_ram_counts[elem] = sum(
+            np.sum(elem_counts[:, :, b], axis=(1, 2)) for b in c.ANTI_RAM_HISTOGRAM_BINS
+        )
 
-        # If there's a gap and we have an active interval, close it
-        if (delta_time > delay_max) & (begin > 0.0):
-            end = met[index - 1]
-            logger.debug(f"    Closing interval due to time gap: {begin} -> {end}")
+    # Pre-compute expected exposure times [s] for the averaging and summing windows.
+    exposure = c.HISTOGRAM_CYCLE_EPOCHS * c.N_CYCLE_AVE * c.EXPOSURE_FACTOR
```
Comment (tmplummer) [inline imap_processing/lo/l1b/lo_l1b.py:2616]: How will hard coded exposure times work on off-nominal pointings?

File: imap_processing/lo/l1b/lo_l1b.py [hunk 5]
```diff
@@ -2515,319 +2511,295 @@ def l1b_bgrates_and_goodtimes(  # noqa: PLR0912
...
-    l1b_backgrounds_and_goodtimes_ds["start_met"] = xr.DataArray(
-        data=bg_start_met,
-        name="start_met",
-        dims=["met"],
-        attrs=attr_mgr_l1b.get_variable_attributes("met"),
+    l1b_combined_ds["epoch"].attrs["DEPEND_0"] = "epoch"
+
+    l1b_combined_ds["pivot"] = xr.DataArray(
+        data=np.float32(pivot),
+        name="pivot",
```
Comment (tmplummer) [inline imap_processing/lo/l1b/lo_l1b.py:2779]: You will need to add this variable to the cdf attribute YAML file and get those attributes here.

File: imap_processing/lo/l1b/lo_l1b.py [hunk 6]
```diff
@@ -2515,319 +2511,295 @@ def l1b_bgrates_and_goodtimes(  # noqa: PLR0912
...
     )
-    l1b_backgrounds_and_goodtimes_ds["end_met"] = xr.DataArray(
-        data=bg_end_met,
-        name="end_met",
-        dims=["met"],
-        attrs=attr_mgr_l1b.get_variable_attributes("met"),
+    l1b_combined_ds["pivot_de"] = xr.DataArray(
+        data=np.float32(pivot_de),
+        name="pivot_de",
```
Comment (tmplummer) [inline imap_processing/lo/l1b/lo_l1b.py:2784]: You will need to add this variable to the cdf attribute YAML file and get those attributes here.

**Review by tmplummer — COMMENTED** [review]
Just one major concern with this implementation. I left some other relevant comments too.

File: imap_processing/lo/constants.py (hunk 3 above)
Comment (vineetbansal) [reply imap_processing/lo/constants.py]: Let me point @nschwadron to this comment - we could put these in a csv but that won't address your concerns. I'm not sure if these overrides are derivable from someplace..

File: imap_processing/lo/l1b/lo_l1b.py (hunk 2 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py:2557]: I'll bring this up with @nschwadron tomorrow..

File: imap_processing/lo/l1b/lo_l1b.py (hunk 4 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py:2616]: Will bring it up with @nschwadron ..

**Comment by tmplummer** [comment]
Does this work have an associated issue? It would be good to keep with best practices of capturing all work in a issue and linking PRs using the "Closes: #nnnn"

File: imap_processing/lo/l1b/lo_l1b.py (hunk 1 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py]: Done.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 6 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py:2784]: Done, for this and other attributes.

**Comment by vineetbansal** [comment]
Opened issue #3135 for this PR specifically.

File: imap_processing/lo/l1b/lo_l1b.py [hunk 7]
```diff
@@ -2558,7 +2559,7 @@ def l1b_bgrates_and_goodtimes(  # noqa: PLR0912
     epoch_utc_str = et_to_utc(ttj2000ns_to_et(epoch_ttj2000[0]))
     epoch_start_dt = datetime.strptime(epoch_utc_str.split("T")[0], "%Y-%m-%d")
     epoch_year = epoch_start_dt.year
-    epoch_doy = epoch_start_dt.timetuple().tm_yday
+    epoch_doy = epoch_to_doy(epoch_ttj2000[:1])[0]
```
Comment (tmplummer) [inline imap_processing/lo/l1b/lo_l1b.py]: Sorry, I missed that you were extracting the year in addition to doy. This implementation is fine, but could be cleaned up by:
```
epoch_start_dt = spiceypy.et2datetime(ttj2000ns_to_et(epoch_ttj2000[0]))
epoch_year = epoch_start_dt.year
epoch_doy = epoch_start_dt.timetuple().tm_yday
```

File: imap_processing/lo/l1b/lo_l1b.py (hunk 2 above)
Comment (frahmanif) [reply imap_processing/lo/l1b/lo_l1b.py:2557]: Good point. The current implementation assumes a nominal long pointing and uses a fixed 3–15 hour window from the beginning of the HK file to estimate the pivot angle. That approach may not be robust for shortened pointings, such as those around TCM activities. Defining the HK selection window dynamically based on the actual pointing duration sounds like a better approach. I would recommend applying only a small edge trim to avoid transition periods.

File: imap_processing/lo/constants.py (hunk 3 above)
Comment (frahmanif) [reply imap_processing/lo/constants.py]: One possible approach could be to move these exception values into an ancillary file rather than hard-coding them in the algorithm.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 4 above)
Comment (frahmanif) [reply imap_processing/lo/l1b/lo_l1b.py:2616]: The exposure values here are tied to the histogram cadence rather than the total pointing duration. The algorithm processes however many histogram cycles are present in the file, while each cycle still corresponds to a fixed 420-second interval. In our current runs, we use a 7-cycle averaging window (~50 minutes) that advances one cycle at a time through the file. So shorter or off-nominal pointings would mainly result in fewer total cycles being analyzed, rather than changing the effective exposure per rate estimate.

File: imap_processing/lo/constants.py (hunk 3 above)
Comment (tmplummer) [reply imap_processing/lo/constants.py]: @frahmanif 
Yes, at the very least, this needs to be in an ancillary file so that it can be updated without making code changes. But, what I was pointing out is that Dave McComas has stipulated that these algorithms should be fully automated, not needing human in the loop changes to tweak things ad hoc. Hi had planned to allow custom goodtimes files to be generated and uploaded by the instrument team. That got shut down because it was not fully automated. That was what I was trying to express with my concern above.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 2 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py:2557]: @tmplummer - my understanding from @frahmanif is that instead of a +3hr and +15hr offset from the start of the hk file, we'll do a +p hr offset from the start and -q hr offset from the end for the window (both constants in `constants.py`. Around an hour to start. I'll implement this change tomorrow, unless you see a problem with this approach

File: imap_processing/lo/constants.py (hunk 3 above)
Comment (vineetbansal) [reply imap_processing/lo/constants.py]: @tmplummer - I'll add these to an ancillary file. I couldn't get an answer from Nathan on how these exceptions came about..

File: imap_processing/lo/l1b/lo_l1b.py (hunk 4 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py:2616]: @tmplummer - I guess this is a non-issue then? Let me know if further clarification is needed from the team.

File: imap_processing/lo/constants.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+# One histogram accumulation cycle duration [s]
+HISTOGRAM_CYCLE_EPOCHS: int = 420
```
Comment (tmplummer) [inline imap_processing/lo/constants.py]: Does it matter that this is just an estimate? I am assuming this is `7 ESA steps * 4 spins/step * 15 seconds/spin`. But the true spin duration is *not* 15 seconds. This is the type of hard coded value that adds lots of technical debt.

It looks like `N_ESA_LEVELS` is defined below, so perhaps this should actually be:
`HISTOGRAM_CYCLE_EPOCHS = N_ESA_LEVELS * N_SPINS_PER_ESA_LEVEL * 15`?

@vineetbansal If this is going to stay as is, please add a TODO and link it to a github issue.

File: imap_processing/lo/constants.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    7,
+)  # ESA levels for RAM estimation (1-indexed)
+
+# Histogram angular bins (0-indexed) corresponding to the RAM and anti-RAM look
+# directions
+RAM_HISTOGRAM_BINS: tuple[slice, ...] = (slice(0, 20), slice(50, 60))
+ANTI_RAM_HISTOGRAM_BINS: tuple[slice, ...] = (slice(20, 50),)
+
+# Nominal background rates [counts/s] for each species
+BG_RATES = {"H": 0.0014925, "O": 0.000136635}
```
Comment (tmplummer) [inline imap_processing/lo/constants.py]: I envision this will also want to be changed in the future.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 2 above)
Comment (tmplummer) [reply imap_processing/lo/l1b/lo_l1b.py:2557]: Ok, that would be an improvement, but will break if the pointing is less than 2 hours long.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 4 above)
Comment (tmplummer) [reply imap_processing/lo/l1b/lo_l1b.py:2616]: Yeah, that makes sense to me now.

File: imap_processing/lo/constants.py (hunk 3 above)
Comment (frahmanif) [reply imap_processing/lo/constants.py]: The days 62, 64, 65 were right before the software issue. I think the exceptions on the background level were implemented to remove those dates. I should still double check with Nathan though.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 2 above)
Comment (frahmanif) [reply imap_processing/lo/l1b/lo_l1b.py:2557]: Pointings shorter than ~2 hours are most likely associated with tests or special operational modes and probably would not need to be included in the nominal goodtimes list. I will double-check this with Nathan, but my understanding is that we should likely define a minimum pointing duration required for a pointing to qualify for the goodtimes analysis.

**Review by tmplummer — APPROVED** [review]
Looks good once the checks pass.

File: imap_processing/lo/constants.py (hunk 8 above)
Comment (vineetbansal) [reply imap_processing/lo/constants.py]: Good catch! I've introduced a `N_SPINS_PER_ESA_LEVEL` and am deriving `HISTOGRAM_CYCLE_EPOCHS` from it now. 2 other existing places that had a hardcoded `4` now refer to `N_SPINS_PER_ESA_LEVEL` instead.

For the spin duration, I've added a `NOMINAL_SPIN_PERIOD_SEC: float = 15.0`, with an explanation that it is not exact. I couldn't find another place in the code where I could grab this "constant". Hope this works for now.

**Comment by vineetbansal** [comment]
@tmplummer - for the time window to get a median pivot angle, @nschwadron mentioned that he'd think about it (his idea was slightly different from what @frahmanif suggested above). I've opened an issue #3169 on this so we can revisit it (shortly in the next couple of days I think).

File: imap_processing/lo/constants.py (hunk 3 above)
Comment (vineetbansal) [reply imap_processing/lo/constants.py]: Pulling this from an ancillary file now.

File: imap_processing/lo/l1b/lo_l1b.py (hunk 7 above)
Comment (vineetbansal) [reply imap_processing/lo/l1b/lo_l1b.py]: Done.
